flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java

```diff
@@ -782,21 +847,57 @@ public KeyGroupRange getKeyGroupRange() {
             }
 
             @Override
+            @SuppressWarnings("unchecked")
             public <N, S extends State, V> S createKeyedState(
                     TypeSerializer<N> namespaceSerializer, StateDescriptor<S, V> stateDescriptor)
                     throws Exception {
-                return ChangelogKeyedStateBackend.this.getOrCreateKeyedState(
-                        namespaceSerializer, stateDescriptor);
+                InternalKvState<K, N, V> kvState =
+                        keyedStateBackend.createOrUpdateInternalState(
+                                namespaceSerializer, stateDescriptor, noTransform(), true);
+                ChangelogState changelogState =
+                        changelogStateFactory.getExistingState(
+                                stateDescriptor.getName(),
+                                StateMetaInfoSnapshot.BackendStateType.KEY_VALUE);
+                if (changelogState == null) {
+                    changelogState =
+                            changelogStateFactory.create(
+                                    stateDescriptor,
+                                    kvState,
+                                    getKvStateChangeLogger(kvState, stateDescriptor, noTransform()),
+                                    keyedStateBackend /* pass the nested backend as key context so that it get key updates on recovery*/);
+                } else {
+                    updateChangelogState(changelogState, kvState, stateDescriptor, noTransform());
+                }
```

Comment:
I'm afraid this branch is not covered by tests.
`ChangelogStateBackendMigrationTest` only has one metadata change and one recovery, so it only triggers `changelogState == null` branch.

Reply:
Sorry for the late reply.
I think current implementation should have covered the case you pointed out, because the `create` would be called when changelog recovery, and the `update` would be called when user access.
You could debug `testKeyedValueStateMigration` for changelog state backend and see the branch has been accessed.